src/xdsl/rewriter.py

```diff
@@ -0,0 +1,199 @@
+import inspect
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from typing import List, Optional, Callable, Dict
+
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import Operation, SSAValue, Region, Block
+
+
+@dataclass(init=False)
+class RewriteAction:
+    """
+    Action that a single rewrite may execute.
+    A rewrite always delete the matched operation, and replace it with new operations.
+    The matched operation results are replaced with new ones.
+    """
+
+    new_ops: List[Operation]
+    """New operations that replace the one matched."""
+
+    new_results: List[SSAValue]
+    """SSA values that replace the matched operation results."""
+    def __init__(self, new_ops: List[Operation],
+                 new_results: Optional[List[SSAValue]]):
+        self.new_ops = new_ops
+        if new_results is not None:
+            self.new_results = new_results
+            return
+
+        if len(new_ops) == 0:
+            self.new_results = []
+        else:
+            self.new_results = new_ops[-1].results
+
+
+class RewritePattern(ABC):
+    """
+    A side-effect free rewrite pattern matching on a DAG.
+    """
+    @abstractmethod
+    def match_and_rewrite(
+            self, op: Operation,
+            new_operands: List[SSAValue]) -> Optional[RewriteAction]:
+        """
+        Match an operation, and optionally returns a rewrite to be performed.
+        `op` is the operation to match, and `new_operands` are the potential new values of the operands.
+        This function returns `None` if the pattern did not matched, and a rewrite action otherwise.
+        """
+        ...
+
+
+@dataclass(eq=False, repr=False)
+class AnonymousRewritePattern(RewritePattern):
+    """
+    A rewrite pattern encoded by an anonymous function.
+    """
+    func: Callable[[Operation, List[SSAValue]], Optional[RewriteAction]]
+
+    def match_and_rewrite(
+            self, op: Operation,
+            new_operands: List[SSAValue]) -> Optional[RewriteAction]:
+        return self.func(op, new_operands)
+
+
+def op_type_rewrite_pattern(func):
+    """
+    This function is intended to be used as a decorator on a RewritePatter method.
+    It uses type hints to match on a specific operation type before calling the decorated function.
+    """
+    # Get the operation argument and check that it is a subclass of Operation
+    params = [param for param in inspect.signature(func).parameters.values()]
+    if len(params) == 3:
```

Comment:
I think that this isn't a safe check. 
We should use `inspect.ismethod()` instead to check if it is actually a method. Afterwards we can check the amount of parameters that then should be either 2 or 3. If that's not the case, we raise an exception.

Reply:
Oh thanks, I didn't knew about that!